src/llama-sampling.cpp

```diff
@@ -465,171 +409,887 @@ void llama_sample_repetition_penalties_impl(
+
+        llama_constraint_softmax_impl(cur_p);
+
+        // Estimate s_hat using the most probable m tokens
+        float s_hat = 0.0;
+        float sum_ti_bi = 0.0;
+        float sum_ti_sq = 0.0;
+        for (size_t i = 0; i < size_t(ctx->m - 1) && i < cur_p->size - 1; ++i) {
+            float t_i = logf(float(i + 2) / float(i + 1));
+            float b_i = logf(cur_p->data[i].p / cur_p->data[i + 1].p);
+            sum_ti_bi += t_i * b_i;
+            sum_ti_sq += t_i * t_i;
+        }
+        s_hat = sum_ti_bi / sum_ti_sq;
+
+        // Compute k from the estimated s_hat and target surprise value
+        float epsilon_hat = s_hat - 1;
+        float k = powf((epsilon_hat * powf(2, ctx->mu)) / (1 - powf(ctx->vocab->n_vocab, -epsilon_hat)), 1 / s_hat);
+
+        llama_constraint_top_k_impl(cur_p, int(k), 1);
+
+        // remember the order to be able to compute the distance later when accepting the token
+        ctx->cur.resize(cur_p->size);
+        for (size_t i = 0; i < cur_p->size; ++i) {
+            ctx->cur[i] = cur_p->data[i];
+        }
+    },
+    /* .reset  = */ [](struct llama_constraint * cnstr) {
+        auto * ctx = (llama_constraint_context_mirostat *) cnstr->ctx;
+        ctx->mu = 0.0f;
+    },
+    /* .copy   = */ [](const struct llama_constraint * cnstr) {
+        const auto * ctx = (const llama_constraint_context_mirostat *) cnstr->ctx;
+        return llama_constraint_init_mirostat_impl(*ctx->vocab, ctx->tau, ctx->eta, ctx->m);
+    },
+    /* .free   = */ [](struct llama_constraint * cnstr) {
+        delete (llama_constraint_context_mirostat *) cnstr->ctx;
+    },
+};
+
+struct llama_constraint * llama_constraint_init_mirostat_impl(
+        const struct llama_vocab & vocab,
+                           float   tau,
+                           float   eta,
+                         int32_t   m) {
+    struct llama_constraint * result = new llama_constraint {
+        /* .iface = */ &llama_constraint_mirostat_i,
+        /* .ctx   = */ new llama_constraint_context_mirostat {
+            /*.vocab =*/ &vocab,
+            /*.tau   =*/ tau,
+            /*.eta   =*/ eta,
+            /*.m     =*/ m,
+            /*.mu    =*/ 0.0f,
+            /*.cur   =*/ {},
+        },
+    };
+
+    return result;
+}
+
+// mirostat v2
+
+struct llama_constraint_context_mirostat_v2 {
+    const float tau;
+    const float eta;
+
+    float mu;
+
+    std::vector<llama_token_data> cur;
+};
+
+static struct llama_constraint_i llama_constraint_mirostat_v2_i = {
+    /* .name   = */ [](const struct llama_constraint * /*cnstr*/) { return "mirostat-v2"; },
+    /* .accept = */ [](struct llama_constraint * cnstr, llama_token token) {
+        auto * ctx = (llama_constraint_context_mirostat_v2 *) cnstr->ctx;
+
+        int32_t idx = -1;
+        for (size_t i = 0; i < ctx->cur.size(); ++i) {
+            if (ctx->cur[i].id == token) {
+                idx = i;
+                break;
+            }
+        }
```

Comment:
I'll try to figure out a way to remove `llama_constraint` and only have `llama_sampler`. The only use case that I think has to be resolved is how to sample greedily, but still have the probabilities of all the candidates.

Reply:
If the first candidate is always selected as the accepted token, then for greedy sampling all you would only need to sort the candidates.

Reply:
Make sense. Regarding the API in `llama.h`, if I removed `llama_constraint` and effectively apply the same abstract interface to `llama_sampler`, I'm not sure how to handle the following calls:

```c
    LLAMA_API void llama_sampler_set_logits(struct llama_sampler * smpl, const float * logits);

    LLAMA_API llama_token_data_array * llama_sampler_get_candidates(struct llama_sampler * smpl);

    LLAMA_API void                      llama_sampler_constraint_add(      struct llama_sampler * smpl, struct llama_constraint * cnstr);
    LLAMA_API int                       llama_sampler_n_constraints (const struct llama_sampler * smpl);
    LLAMA_API struct llama_constraint * llama_sampler_constraint_get(const struct llama_sampler * smpl, int32_t i);

    LLAMA_API llama_token llama_sampler_sample(struct llama_sampler * smpl, llama_token_data_array * cur_p);

    LLAMA_API int llama_sampler_n_prev(const struct llama_sampler * smpl);

    LLAMA_API llama_token llama_sampler_prev(const struct llama_sampler * smpl, int32_t ith);

    LLAMA_API llama_token llama_sampler_last(const struct llama_sampler * smpl);
```

These calls make sense only for the "main" sampler - the one that we attach all sub-samplers to. For example:

```c
llama_sampler * top_k = llama_sampler_init_top_k(k);

...

// these should fail or be no-op:
llama_sampler_set_logits(top_k, logits);
llama_sampler_sample(top_k, cur_p);
llama_sampler_prev(top_k);
```

They should work only with the "main" sampler:

```c
llama_sampler_params sparams = llama_sampler_default_params();
llama_sampler * smpl = llama_sampler_init(model, sparams);

...

// these should work:
llama_sampler_set_logits(smpl, logits);
llama_sampler_sample(smpl, cur_p);
llama_sampler_prev(smpl);
```

What would be a good way to do this? I can think of 2 options, thought neither seems great:

- extend the abstract interface with all these functions and have them `null` for the regular samplers
- check the `name` of the sampler to determine if it is the "main" one

Any suggestions?

Reply:
I do not think that these functions should be added to what currently are the constraints objects (I will refer to them as samplers from this point). My initial idea to create chains of samplers was to have a type of sampler that internally contains a list of other samplers that it calls in order. So for example:

```C++
{
    llama_sampler_t sampler = llama_sampler_chain_new();
    llama_sampler_chain_add(sampler, llama_sampler_top_k_new(10));
    llama_sampler_chain_add(sampler, llama_sampler_temperature_new(0.5));
    llama_sampler_chain_add(sampler, llama_sampler_top_p_new(0.9));
    llama_sampler_chain_add(sampler, llama_sampler_softmax_new());
    llama_sampler_chain_add(sampler, llama_sampler_sample_dist_new(seed)); // contains the RNG state
    // llama_sampler_chain_add(sampler, llama_sampler_sample_greedy_new());

    token = llama_sampler_sample(llama_ctx, sampler);
}

// basic llama_sampler_sample implementation
llama_token llama_sampler_sample(llama_context * ctx, llama_sampler * sampler, int ith) {
    // construct the sampling state from the logits
    float * logits = llama_get_logits_ith(ctx, ith);
    llama_token_data_array candidates = { /* initialize from logits */ };

    // run the sampler
    sampler->sample(ctx, candidates);

    // get a token
    llama_token token = candidates.data[0].id; // or candidates.data[candidates.best_idx].id;

    sampler->accept(token);

    return token;
}
```

I guess this would still be missing a way to obtain the previously sampled tokens, but I think that should probably be done in llama.cpp by associating KV cells with a token and adding functions to return the tokens of each sequence.

Reply:
Ok, I think it is clear now. Will give this another go.